python/pyarrow/_dataset.pyx

```diff
@@ -0,0 +1,1019 @@
+        if context is None:
+            result = self.dataset.NewScan()
+        else:
+            result = self.dataset.NewScanWithContext(context.unwrap())
+        return ScannerBuilder.wrap(GetResultValue(result))
+
+    @property
+    def sources(self):
+        cdef vector[shared_ptr[CDataSource]] sources = self.dataset.sources()
+        return [DataSource.wrap(source) for source in sources]
+
+    @property
+    def schema(self):
+        return pyarrow_wrap_schema(self.dataset.schema())
+
+
+cdef class ScanContext:
+
+    cdef:
+        shared_ptr[CScanContext] wrapped
+        CScanContext *context
+
+    def __init__(self, MemoryPool memory_pool=None):
+        cdef shared_ptr[CScanContext] context = make_shared[CScanContext]()
+        self.init(context)
+        if memory_pool is not None:
+            self.memory_pool = memory_pool
+
+    cdef init(self, shared_ptr[CScanContext]& sp):
+        self.wrapped = sp
+        self.context = sp.get()
+
+    @staticmethod
+    cdef ScanContext wrap(shared_ptr[CScanContext]& sp):
+        cdef ScanContext self = ScanContext.__new__(ScanContext)
+        self.init(sp)
+        return self
+
+    cdef inline shared_ptr[CScanContext] unwrap(self):
+        return self.wrapped
+
+    @property
+    def memory_pool(self):
+        # TODO(kszucs): there's no function to box/wrap memory pool
+        cdef:
+            MemoryPool pool = MemoryPool.__new__(MemoryPool)
+        # this might be unsafe because of the raw pointers
+        pool.init(self.context.pool)
+        return pool
+
+    @memory_pool.setter
+    def memory_pool(self, MemoryPool pool):
+        self.context.pool = maybe_unbox_memory_pool(pool)
+
+
+cdef class ScanTask:
+    """Read record batches from a range of a single data fragment.
+
+    A ScanTask is meant to be a unit of work to be dispatched.
+    """
+
+    cdef:
+        shared_ptr[CScanTask] wrapped
+        CScanTask* task
+
+    def __init__(self):
+        _forbid_instantiation(self.__class__, subclasses_instead=False)
+
+    cdef init(self, shared_ptr[CScanTask]& sp):
+        self.wrapped = sp
+        self.task = self.wrapped.get()
+
+    @staticmethod
+    cdef wrap(shared_ptr[CScanTask]& sp):
+        cdef SimpleScanTask self = SimpleScanTask.__new__(SimpleScanTask)
+        self.init(sp)
+        return self
+
+    cdef inline shared_ptr[CScanTask] unwrap(self):
+        return self.wrapped
+
+    def scan(self):
+        """Iterate through sequence of materialized record batches.
+
+        Execution semantics are encapsulated in the particular ScanTask
+        implementation.
+
+        Returns
+        -------
+        record_batches : iterator of RecordBatch
+        """
+        cdef:
+            CRecordBatchIterator iterator
+            shared_ptr[CRecordBatch] record_batch
+
+        iterator = move(GetResultValue(move(self.task.Scan())))
+
+        while True:
+            iterator.Next(&record_batch)
```

Comment:
Unchecked status? (do we have a CI job that will find unchecked status in Cython)